### ISSUE TITLE
rfe: async handler API

### ISSUE THREAD

**Issue by agrover** [agrover/tcmu-runner#12]
We should enable those handlers that wish to, to complete requests out of order.

**Comment by os12** [comment]
Right, this needed for serious embedding. That is, I would like to take the request processing code from tcmu-runner (see #13), link it to an external application (as it manages its own threads) and then process the requests in an async (ie non-blocking) way. Note, the notification itself should work with `epoll()` today, yet the request processing is still blocking.

Looking at `tcmu_handle_device_events()`/`handle_one_command()`, the plugin should be able to complete/fail a request as well as mark it as "pending". It should then be possible to complete a pending request (possibly out of order).

I can take care of the user-mode work, including any and all API/embedding extensions and issues. Yet, this feature would require kernel support as the behavior of the circular queue would change. Also, it would be good to map the individual iovecs for each request, rather than copy the data back and forth.

**Comment by os12** [comment]
Hi Andy, the recent work from @mkhon submitted in #21 is for this issue. Maxim extended the Runner's "file" handler with async options, but left the "consumer" example untouched. I've just had a look - it should be trivial to correct/extend. One of us will look into that.

Could you clarify the handler binding rules please:
  1. Can multiple consumer-like processes bind to distinct TCMU devices?
  1. What is the best way to implement device selection within each process? It looks like I must implement the `added` handler to accept/skip based on the value of `tcmu_get_dev_cfgstring(dev)`.

Thanks!

**Comment by agrover** [comment]
Hi Oleg,

1. the idea was that different, unrelated, processes could bind to distinct TCMU devices, as differentiated by part of the cfgstring called "subtype". 
1. each process should only get added() for the subtype it registered as supporting, so it should not need to skip or accept. It will likely need the remainder of the cfgstring to know how to configure itself to handle that particular device.

**Comment by os12** [comment]
Right, I can easily put a discriminant into subtype. Thanks! (I plan to 
have multiple devices of a similar type handled by distinct processes 
for stability reasons. So, the daemons need to choose the device.)

**Comment by os12** [comment]
Update:
  * The code landed in #26 (from @mkhon)
  * There is a notification issue between the kernel- and user-mode pieces. A PR is coming...
  * The kernel-mode code needs quite a bit of work:
    * The first cut of the implementation is here: https://github.com/rancher/linux/tree/v4.2-tcmu-async
    * There are issues with ring sizes, we are cleaning that up
